### Module 1 Coding Questions

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
import datetime

%matplotlib inline
sns.set_style('dark')
sns.set(font_scale=1.2)

import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns',None)
#pd.set_option('display.max_rows',None)

np.random.seed(0)
np.set_printoptions(suppress=True)

Retrieve all the records from the Employees table

In [2]:
df = pd.read_csv("Employee.csv")

In [3]:
df

EmployeeId  LastName FirstName                Title  ReportsTo  \
0           1     Adams    Andrew      General Manager        NaN   
1           2   Edwards     Nancy        Sales Manager        1.0   
2           3   Peacock      Jane  Sales Support Agent        2.0   
3           4      Park  Margaret  Sales Support Agent        2.0   
4           5   Johnson     Steve  Sales Support Agent        2.0   
5           6  Mitchell   Michael           IT Manager        1.0   
6           7      King    Robert             IT Staff        6.0   
7           8  Callahan     Laura             IT Staff        6.0   

             BirthDate             HireDate                      Address  \
0  1962-02-18 00:00:00  2002-08-14 00:00:00          11120 Jasper Ave NW   
1  1958-12-08 00:00:00  2002-05-01 00:00:00                 825 8 Ave SW   
2  1973-08-29 00:00:00  2002-04-01 00:00:00                1111 6 Ave SW   
3  1947-09-19 00:00:00  2003-05-03 00:00:00             683 10 Street SW   
4  1965-03-03 00:00:00  2003-10-17 00:00:00                 7727B 41 Ave   
5  1973-07-01 00:00:00  2003-10-17 00:00:00         5827 Bowness Road NW   
6  1970-05-29 00:00:00  2004-01-02 00:00:00  590 Columbia Boulevard West   
7  1968-01-09 00:00:00  2004-03-04 00:00:00                  923 7 ST NW   

         City State Country PostalCode              Phone                Fax  \
0    Edmonton    AB  Canada    T5K 2N1  +1 (780) 428-9482  +1 (780) 428-3457   
1     Calgary    AB  Canada    T2P 2T3  +1 (403) 262-3443  +1 (403) 262-3322   
2     Calgary    AB  Canada    T2P 5M5  +1 (403) 262-3443  +1 (403) 262-6712   
3     Calgary    AB  Canada    T2P 5G3  +1 (403) 263-4423  +1 (403) 263-4289   
4     Calgary    AB  Canada    T3B 1Y7   1 (780) 836-9987   1 (780) 836-9543   
5     Calgary    AB  Canada    T3B 0C5  +1 (403) 246-9887  +1 (403) 246-9899   
6  Lethbridge    AB  Canada    T1K 5N8  +1 (403) 456-9986  +1 (403) 456-8485   
7  Lethbridge    AB  Canada    T1H 1Y8  +1 (403) 467-3351  +1 (403) 467-8772   

                      Email  
0    andrew@chinookcorp.com  
1     nancy@chinookcorp.com  
2      jane@chinookcorp.com  
3  margaret@chinookcorp.com  
4     steve@chinookcorp.com  
5   michael@chinookcorp.com  
6    robert@chinookcorp.com  
7     laura@chinookcorp.com

What is Robert King's mailing address? Note: You will have to scroll to the right in order to see it.

590 Columbia Boulevard West

Retrieve the FirstName, LastName, Birthdate, Address, City, and State from the Employees table

In [4]:
emp = df[['FirstName','LastName', 'BirthDate','Address', 'City', 'State']]

In [5]:
emp

FirstName  LastName            BirthDate                      Address  \
0    Andrew     Adams  1962-02-18 00:00:00          11120 Jasper Ave NW   
1     Nancy   Edwards  1958-12-08 00:00:00                 825 8 Ave SW   
2      Jane   Peacock  1973-08-29 00:00:00                1111 6 Ave SW   
3  Margaret      Park  1947-09-19 00:00:00             683 10 Street SW   
4     Steve   Johnson  1965-03-03 00:00:00                 7727B 41 Ave   
5   Michael  Mitchell  1973-07-01 00:00:00         5827 Bowness Road NW   
6    Robert      King  1970-05-29 00:00:00  590 Columbia Boulevard West   
7     Laura  Callahan  1968-01-09 00:00:00                  923 7 ST NW   

         City State  
0    Edmonton    AB  
1     Calgary    AB  
2     Calgary    AB  
3     Calgary    AB  
4     Calgary    AB  
5     Calgary    AB  
6  Lethbridge    AB  
7  Lethbridge    AB

Which of the employees listed below has a birthdate of 3-3-1965?

Steve Johnson

Retrieve all the columns from the Tracks table, but only return 20 rows

In [6]:
tracks = pd.read_csv('Track.csv')

In [7]:
tracks.head(20)

TrackId                                     Name  AlbumId  MediaTypeId  \
0         1  For Those About To Rock (We Salute You)        1            1   
1         2                        Balls to the Wall        2            2   
2         3                          Fast As a Shark        3            2   
3         4                        Restless and Wild        3            2   
4         5                     Princess of the Dawn        3            2   
5         6                    Put The Finger On You        1            1   
6         7                          Let's Get It Up        1            1   
7         8                         Inject The Venom        1            1   
8         9                               Snowballed        1            1   
9        10                               Evil Walks        1            1   
10       11                                   C.O.D.        1            1   
11       12                       Breaking The Rules        1            1   
12       13                 Night Of The Long Knives        1            1   
13       14                               Spellbound        1            1   
14       15                                  Go Down        4            1   
15       16                              Dog Eat Dog        4            1   
16       17                        Let There Be Rock        4            1   
17       18                           Bad Boy Boogie        4            1   
18       19                            Problem Child        4            1   
19       20                                 Overdose        4            1   

    GenreId                                           Composer  Milliseconds  \
0         1          Angus Young, Malcolm Young, Brian Johnson        343719   
1         1                                                NaN        342562   
2         1  F. Baltes, S. Kaufman, U. Dirkscneider & W. Ho...        230619   
3         1  F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. D...        252051   
4         1                         Deaffy & R.A. Smith-Diesel        375418   
5         1          Angus Young, Malcolm Young, Brian Johnson        205662   
6         1          Angus Young, Malcolm Young, Brian Johnson        233926   
7         1          Angus Young, Malcolm Young, Brian Johnson        210834   
8         1          Angus Young, Malcolm Young, Brian Johnson        203102   
9         1          Angus Young, Malcolm Young, Brian Johnson        263497   
10        1          Angus Young, Malcolm Young, Brian Johnson        199836   
11        1          Angus Young, Malcolm Young, Brian Johnson        263288   
12        1          Angus Young, Malcolm Young, Brian Johnson        205688   
13        1          Angus Young, Malcolm Young, Brian Johnson        270863   
14        1                                              AC/DC        331180   
15        1                                              AC/DC        215196   
16        1                                              AC/DC        366654   
17        1                                              AC/DC        267728   
18        1                                              AC/DC        325041   
19        1                                              AC/DC        369319   

       Bytes  UnitPrice  
0   11170334       0.99  
1    5510424       0.99  
2    3990994       0.99  
3    4331779       0.99  
4    6290521       0.99  
5    6713451       0.99  
6    7636561       0.99  
7    6852860       0.99  
8    6599424       0.99  
9    8611245       0.99  
10   6566314       0.99  
11   8596840       0.99  
12   6706347       0.99  
13   8817038       0.99  
14  10847611       0.99  
15   7032162       0.99  
16  12021261       0.99  
17   8776140       0.99  
18  10617116       0.99  
19  12066294       0.99

What is the runtime in milliseconds for the 5th track, entitled "Princess of the Dawn"? 

375418

### Module 2 Coding Assignment

Find all the tracks that have a length of 5,000,000 milliseconds or more

In [8]:
trackslong = tracks[tracks['Milliseconds'] >= 5000000]

In [9]:
trackslong

TrackId                     Name  AlbumId  MediaTypeId  GenreId  \
2819     2820   Occupation / Precipice      227            3       19   
3223     3224  Through a Looking Glass      229            3       21   

     Composer  Milliseconds       Bytes  UnitPrice  
2819      NaN       5286953  1054423946       1.99  
3223      NaN       5088838  1059546140       1.99

How many tracks are returned?

2

Find all the invoices whose total is between $5 and $15 dollars

In [10]:
invoice = pd.read_csv("Invoice.csv")

In [11]:
invoice

InvoiceId  CustomerId          InvoiceDate  \
0            1           2  2009-01-01 00:00:00   
1            2           4  2009-01-02 00:00:00   
2            3           8  2009-01-03 00:00:00   
3            4          14  2009-01-06 00:00:00   
4            5          23  2009-01-11 00:00:00   
..         ...         ...                  ...   
407        408          25  2013-12-05 00:00:00   
408        409          29  2013-12-06 00:00:00   
409        410          35  2013-12-09 00:00:00   
410        411          44  2013-12-14 00:00:00   
411        412          58  2013-12-22 00:00:00   

                               BillingAddress BillingCity BillingState  \
0                     Theodor-Heuss-Straße 34   Stuttgart          NaN   
1                            Ullevålsveien 14        Oslo          NaN   
2                             Grétrystraat 63    Brussels          NaN   
3                              8210 111 ST NW    Edmonton           AB   
4                             69 Salem Street      Boston           MA   
..                                        ...         ...          ...   
407                     319 N. Frances Street     Madison           WI   
408                    796 Dundas Street West     Toronto           ON   
409  Rua dos Campeões Europeus de Viena, 4350       Porto          NaN   
410                           Porthaninkatu 9    Helsinki          NaN   
411                       12,Community Centre       Delhi          NaN   

    BillingCountry BillingPostalCode  Total  
0          Germany             70174   1.98  
1           Norway              0171   3.96  
2          Belgium              1000   5.94  
3           Canada           T6G 2C7   8.91  
4              USA              2113  13.86  
..             ...               ...    ...  
407            USA             53703   3.96  
408         Canada           M6J 1V1   5.94  
409       Portugal               NaN   8.91  
410        Finland             00530  13.86  
411          India            110017   1.99  

[412 rows x 9 columns]

In [12]:
invoicebetween = invoice[(invoice["Total"] >= 5.00) & (invoice["Total"] <= 15.00)]

In [13]:
invoicebetween

InvoiceId  CustomerId          InvoiceDate  \
2            3           8  2009-01-03 00:00:00   
3            4          14  2009-01-06 00:00:00   
4            5          23  2009-01-11 00:00:00   
9           10          46  2009-02-03 00:00:00   
10          11          52  2009-02-06 00:00:00   
..         ...         ...                  ...   
401        402          50  2013-11-05 00:00:00   
402        403          56  2013-11-08 00:00:00   
408        409          29  2013-12-06 00:00:00   
409        410          35  2013-12-09 00:00:00   
410        411          44  2013-12-14 00:00:00   

                               BillingAddress   BillingCity BillingState  \
2                             Grétrystraat 63      Brussels          NaN   
3                              8210 111 ST NW      Edmonton           AB   
4                             69 Salem Street        Boston           MA   
9                            3 Chatham Street        Dublin       Dublin   
10                          202 Hoxton Street        London          NaN   
..                                        ...           ...          ...   
401                        C/ San Bernardo 85        Madrid          NaN   
402                        307 Macacha Güemes  Buenos Aires          NaN   
408                    796 Dundas Street West       Toronto           ON   
409  Rua dos Campeões Europeus de Viena, 4350         Porto          NaN   
410                           Porthaninkatu 9      Helsinki          NaN   

     BillingCountry BillingPostalCode  Total  
2           Belgium              1000   5.94  
3            Canada           T6G 2C7   8.91  
4               USA              2113  13.86  
9           Ireland               NaN   5.94  
10   United Kingdom            N1 5LH   8.91  
..              ...               ...    ...  
401           Spain             28015   5.94  
402       Argentina              1106   8.91  
408          Canada           M6J 1V1   5.94  
409        Portugal               NaN   8.91  
410         Finland             00530  13.86  

[168 rows x 9 columns]

While the query in this example is limited to 10 records, running the query correctly will indicate how many total records there are - enter that number below.

168

Find all the customers from the following States: RJ, DF, AB, BC, CA, WA, NY.

In [14]:
customers = pd.read_csv("Customer.csv")

In [15]:
customers

CustomerId  FirstName      LastName  \
0            1       Luís     Gonçalves   
1            2     Leonie        Köhler   
2            3   François      Tremblay   
3            4      Bjørn        Hansen   
4            5  František   Wichterlová   
5            6     Helena          Holý   
6            7     Astrid        Gruber   
7            8       Daan       Peeters   
8            9       Kara       Nielsen   
9           10    Eduardo       Martins   
10          11  Alexandre         Rocha   
11          12    Roberto       Almeida   
12          13   Fernanda         Ramos   
13          14       Mark       Philips   
14          15   Jennifer      Peterson   
15          16      Frank        Harris   
16          17       Jack         Smith   
17          18   Michelle        Brooks   
18          19        Tim         Goyer   
19          20        Dan        Miller   
20          21      Kathy         Chase   
21          22    Heather       Leacock   
22          23       John        Gordon   
23          24      Frank       Ralston   
24          25     Victor       Stevens   
25          26    Richard    Cunningham   
26          27    Patrick          Gray   
27          28      Julia       Barnett   
28          29     Robert         Brown   
29          30     Edward       Francis   
30          31     Martha          Silk   
31          32      Aaron      Mitchell   
32          33      Ellie      Sullivan   
33          34       João     Fernandes   
34          35   Madalena       Sampaio   
35          36     Hannah     Schneider   
36          37       Fynn    Zimmermann   
37          38     Niklas      Schröder   
38          39    Camille       Bernard   
39          40  Dominique      Lefebvre   
40          41       Marc        Dubois   
41          42      Wyatt        Girard   
42          43   Isabelle       Mercier   
43          44      Terhi    Hämäläinen   
44          45   Ladislav        Kovács   
45          46       Hugh      O'Reilly   
46          47      Lucas       Mancini   
47          48   Johannes  Van der Berg   
48          49  Stanisław        Wójcik   
49          50    Enrique         Muñoz   
50          51     Joakim     Johansson   
51          52       Emma         Jones   
52          53       Phil        Hughes   
53          54      Steve        Murray   
54          55       Mark        Taylor   
55          56      Diego     Gutiérrez   
56          57       Luis         Rojas   
57          58      Manoj        Pareek   
58          59       Puja    Srivastava   

                                             Company  \
0   Embraer - Empresa Brasileira de Aeronáutica S.A.   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                   JetBrains s.r.o.   
5                                                NaN   
6                                                NaN   
7                                                NaN   
8                                                NaN   
9                                   Woodstock Discos   
10                              Banco do Brasil S.A.   
11                                            Riotur   
12                                               NaN   
13                                             Telus   
14                                     Rogers Canada   
15                                       Google Inc.   
16                             Microsoft Corporation   
17                                               NaN   
18                                        Apple Inc.   
19                                               NaN   
20                                               NaN   
21                                               NaN   
22                                               NaN   
23                                               NaN   
24                     

In [16]:
customers["State"].value_counts()

SP        3
CA        3
ON        2
NS        1
RM        1
BC        1
Dublin    1
AB        1
WA        1
IL        1
MB        1
AZ        1
NY        1
DF        1
FL        1
RJ        1
NSW       1
TX        1
VV        1
NT        1
NV        1
MA        1
WI        1
QC        1
UT        1
Name: State, dtype: int64

In [17]:
customers[customers["State"] == 'RJ']

CustomerId FirstName LastName Company           Address            City  \
11          12   Roberto  Almeida  Riotur  Praça Pio X, 119  Rio de Janeiro   

   State Country PostalCode               Phone                 Fax  \
11    RJ  Brazil  20040-020  +55 (21) 2271-7000  +55 (21) 2271-7070   

                            Email  SupportRepId  
11  roberto.almeida@riotur.gov.br             3

In [18]:
customers[customers["State"] == 'DF']

CustomerId FirstName LastName Company       Address      City State  \
12          13  Fernanda    Ramos     NaN  Qe 7 Bloco G  Brasília    DF   

   Country PostalCode               Phone                 Fax  \
12  Brazil  71020-677  +55 (61) 3363-5547  +55 (61) 3363-7855   

                       Email  SupportRepId  
12  fernadaramos4@uol.com.br             4

In [19]:
customers[customers["State"] == 'AB']

CustomerId FirstName LastName Company         Address      City State  \
13          14      Mark  Philips   Telus  8210 111 ST NW  Edmonton    AB   

   Country PostalCode              Phone                Fax  \
13  Canada    T6G 2C7  +1 (780) 434-4554  +1 (780) 434-5565   

                 Email  SupportRepId  
13  mphilips12@shaw.ca             5

In [20]:
customers[customers["State"] == 'BC']

CustomerId FirstName  LastName        Company              Address  \
14          15  Jennifer  Peterson  Rogers Canada  700 W Pender Street   

         City State Country PostalCode              Phone                Fax  \
14  Vancouver    BC  Canada    V6C 1G8  +1 (604) 688-2255  +1 (604) 688-8756   

                  Email  SupportRepId  
14  jenniferp@rogers.ca             3

In [21]:
customers[customers["State"] == 'CA']

CustomerId FirstName LastName      Company                    Address  \
15          16     Frank   Harris  Google Inc.  1600 Amphitheatre Parkway   
18          19       Tim    Goyer   Apple Inc.            1 Infinite Loop   
19          20       Dan   Miller          NaN       541 Del Medio Avenue   

             City State Country  PostalCode              Phone  \
15  Mountain View    CA     USA  94043-1351  +1 (650) 253-0000   
18      Cupertino    CA     USA       95014  +1 (408) 996-1010   
19  Mountain View    CA     USA   94040-111  +1 (650) 644-3358   

                  Fax                Email  SupportRepId  
15  +1 (650) 253-0000   fharris@google.com             4  
18  +1 (408) 996-1011     tgoyer@apple.com             3  
19                NaN  dmiller@comcast.com             4

In [22]:
customers[customers["State"] == 'WA']

CustomerId FirstName LastName                Company          Address  \
16          17      Jack    Smith  Microsoft Corporation  1 Microsoft Way   

       City State Country  PostalCode              Phone                Fax  \
16  Redmond    WA     USA  98052-8300  +1 (425) 882-8080  +1 (425) 882-8081   

                      Email  SupportRepId  
16  jacksmith@microsoft.com             5

In [23]:
customers[customers["State"] == 'NY']

CustomerId FirstName LastName Company       Address      City State  \
17          18  Michelle   Brooks     NaN  627 Broadway  New York    NY   

   Country  PostalCode              Phone                Fax  \
17     USA  10012-2612  +1 (212) 221-3546  +1 (212) 221-4679   

                Email  SupportRepId  
17  michelleb@aol.com             3

What company does Jack Smith work for?

Microsoft Corp

In [24]:
invoice

InvoiceId  CustomerId          InvoiceDate  \
0            1           2  2009-01-01 00:00:00   
1            2           4  2009-01-02 00:00:00   
2            3           8  2009-01-03 00:00:00   
3            4          14  2009-01-06 00:00:00   
4            5          23  2009-01-11 00:00:00   
..         ...         ...                  ...   
407        408          25  2013-12-05 00:00:00   
408        409          29  2013-12-06 00:00:00   
409        410          35  2013-12-09 00:00:00   
410        411          44  2013-12-14 00:00:00   
411        412          58  2013-12-22 00:00:00   

                               BillingAddress BillingCity BillingState  \
0                     Theodor-Heuss-Straße 34   Stuttgart          NaN   
1                            Ullevålsveien 14        Oslo          NaN   
2                             Grétrystraat 63    Brussels          NaN   
3                              8210 111 ST NW    Edmonton           AB   
4                             69 Salem Street      Boston           MA   
..                                        ...         ...          ...   
407                     319 N. Frances Street     Madison           WI   
408                    796 Dundas Street West     Toronto           ON   
409  Rua dos Campeões Europeus de Viena, 4350       Porto          NaN   
410                           Porthaninkatu 9    Helsinki          NaN   
411                       12,Community Centre       Delhi          NaN   

    BillingCountry BillingPostalCode  Total  
0          Germany             70174   1.98  
1           Norway              0171   3.96  
2          Belgium              1000   5.94  
3           Canada           T6G 2C7   8.91  
4              USA              2113  13.86  
..             ...               ...    ...  
407            USA             53703   3.96  
408         Canada           M6J 1V1   5.94  
409       Portugal               NaN   8.91  
410        Finland             00530  13.86  
411          India            110017   1.99  

[412 rows x 9 columns]

Find all the invoices for customer 56 and 58 where the total was between $1.00 and $5.00

In [25]:
invoice[invoice["CustomerId"] == 56]

InvoiceId  CustomerId          InvoiceDate      BillingAddress  \
118        119          56  2010-06-12 00:00:00  307 Macacha Güemes   
141        142          56  2010-09-14 00:00:00  307 Macacha Güemes   
163        164          56  2010-12-17 00:00:00  307 Macacha Güemes   
215        216          56  2011-08-07 00:00:00  307 Macacha Güemes   
336        337          56  2013-01-28 00:00:00  307 Macacha Güemes   
347        348          56  2013-03-10 00:00:00  307 Macacha Güemes   
402        403          56  2013-11-08 00:00:00  307 Macacha Güemes   

      BillingCity BillingState BillingCountry BillingPostalCode  Total  
118  Buenos Aires          NaN      Argentina              1106   1.98  
141  Buenos Aires          NaN      Argentina              1106   3.96  
163  Buenos Aires          NaN      Argentina              1106   5.94  
215  Buenos Aires          NaN      Argentina              1106   0.99  
336  Buenos Aires          NaN      Argentina              1106   1.98  
347  Buenos Aires          NaN      Argentina              1106  13.86  
402  Buenos Aires          NaN      Argentina              1106   8.91

In [26]:
invoice[invoice["CustomerId"] == 58]

InvoiceId  CustomerId          InvoiceDate       BillingAddress  \
119        120          58  2010-06-12 00:00:00  12,Community Centre   
130        131          58  2010-07-23 00:00:00  12,Community Centre   
185        186          58  2011-03-23 00:00:00  12,Community Centre   
314        315          58  2012-10-27 00:00:00  12,Community Centre   
337        338          58  2013-01-29 00:00:00  12,Community Centre   
359        360          58  2013-05-03 00:00:00  12,Community Centre   
411        412          58  2013-12-22 00:00:00  12,Community Centre   

    BillingCity BillingState BillingCountry BillingPostalCode  Total  
119       Delhi          NaN          India            110017   1.98  
130       Delhi          NaN          India            110017  13.86  
185       Delhi          NaN          India            110017   8.91  
314       Delhi          NaN          India            110017   1.98  
337       Delhi          NaN          India            110017   3.96  
359       Delhi          NaN          India            110017   5.94  
411       Delhi          NaN          India            110017   1.99

What was the invoice date for invoice ID 315?

In [27]:
invoice[invoice["InvoiceId"] == 315]

InvoiceId  CustomerId          InvoiceDate       BillingAddress  \
314        315          58  2012-10-27 00:00:00  12,Community Centre   

    BillingCity BillingState BillingCountry BillingPostalCode  Total  
314       Delhi          NaN          India            110017   1.98

Find all the tracks whose name starts with 'All'.

In [28]:
tracks

TrackId                                               Name  AlbumId  \
0           1            For Those About To Rock (We Salute You)        1   
1           2                                  Balls to the Wall        2   
2           3                                    Fast As a Shark        3   
3           4                                  Restless and Wild        3   
4           5                               Princess of the Dawn        3   
...       ...                                                ...      ...   
3498     3499  Pini Di Roma (Pinien Von Rom) \ I Pini Della V...      343   
3499     3500  String Quartet No. 12 in C Minor, D. 703 "Quar...      344   
3500     3501               L'orfeo, Act 3, Sinfonia (Orchestra)      345   
3501     3502  Quintet for Horn, Violin, 2 Violas, and Cello ...      346   
3502     3503                                      Koyaanisqatsi      347   

      MediaTypeId  GenreId                                           Composer  \
0               1        1          Angus Young, Malcolm Young, Brian Johnson   
1               2        1                                                NaN   
2               2        1  F. Baltes, S. Kaufman, U. Dirkscneider & W. Ho...   
3               2        1  F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. D...   
4               2        1                         Deaffy & R.A. Smith-Diesel   
...           ...      ...                                                ...   
3498            2       24                                                NaN   
3499            2       24                                     Franz Schubert   
3500            2       24                                 Claudio Monteverdi   
3501            2       24                            Wolfgang Amadeus Mozart   
3502            2       10                                       Philip Glass   

      Milliseconds     Bytes  UnitPrice  
0           343719  11170334       0.99  
1           342562   5510424       0.99  
2           230619   3990994       0.99  
3           252051   4331779       0.99  
4           375418   6290521       0.99  
...            ...       ...        ...  
3498        286741   4718950       0.99  
3499        139200   2283131       0.99  
3500         66639   1189062       0.99  
3501        221331   3665114       0.99  
3502        206005   3305164       0.99  

[3503 rows x 9 columns]

In [29]:
tracks[tracks["Name"].str.contains("All")]

TrackId                                               Name  AlbumId  \
37         38                                  All I Really Want        6   
133       134                                        All For You       14   
384       385                                           All Star       33   
421       422                                      I Want It All       36   
451       452                             Rock And Roll All Nite       37   
802       803                                  Love Conquers All       64   
1008     1009                                        All My Life       81   
1010     1011                                        Have It All       81   
1101     1102                       Doce De Carnaval (Candy All)       87   
1528     1529                                Always Be All Right      123   
1575     1576                             Rock And Roll All Nite      126   
1607     1608                                        All My Love      130   
1782     1783               Gonna Give Her All The Love I've Got      146   
1891     1892                                All Within My Hands      155   
1893     1894                             ...And Justice For All      156   
2191     2192                                        All or None      180   
2273     2274                                 All Dead, All Dead      186   
2508     2509                                     Hands All Over      203   
2830     2831               Taking a Break from All Your Worries      227   
2887     2888             All the Best Cowboys Have Daddy Issues      230   
2968     2969                                 All Because Of You      235   
2990     2991                           All Along The Watchtower      237   
3002     3003                                  All I Want Is You      237   
3016     3017                                  All I Want Is You      238   
3051     3052                Where Have All The Good Times Gone?      242   
3174     3175                                       The Alliance      249   
3315     3316                                        All My Love      258   
3358     3359  Symphony No. 3 in E-flat major, Op. 55, "Eroic...      268   
3373     3374                                    All Night Thing      269   
3405     3406  Concerto No. 1 in E Major, RV 269 "Spring": I....      275   
3411     3412  "Eine Kleine Nachtmusik" Serenade In G, K. 525...      281   
3414     3415      Symphony No.5 in C Minor: I. Allegro con brio      284   
3430     3431  Symphony No.1 in D Major, Op.25 "Classical", A...      298   
3432     3433      Concerto No.2 in F Major, BWV1047, I. Allegro      300   
3488     3489                Symphony No. 2: III. Allegro vivace      334   
3493     3494  Symphony No. 2, Op. 16 -  "The Four Temperamen...      338   
3497     3498  Concerto for Violin, Strings and Continuo in G...      342   
3499     3500  String Quartet No. 12 in C Minor, D. 703 "Quar...      344   
3501     3502  Quintet for Horn, Violin, 2 Violas, and Cello ...      346   

      MediaTypeId  GenreId                                          Composer  \
37              1        1                 Alanis Morissette & Glenn Ballard   
133             1        3                                               NaN   
384             1        7                                        Nando Reis   
421             1        1                                             Queen   
451             1        1                        Paul Stanley, Gene Simmons   
802             1        1                         Blackmore, Glover, Turner   
1008            1        4                                      Foo Fighters   
1010            1        4                                      Foo Fighters   
1101            1        2                                               NaN   
1528            1        7                                               NaN   
1575            1        1                   

While only 10 records are shown, the query will indicate how many total records there are for this query - enter that number below

In [30]:
tracks[tracks["Name"].str.contains("All")].count()

TrackId         39
Name            39
AlbumId         39
MediaTypeId     39
GenreId         39
Composer        32
Milliseconds    39
Bytes           39
UnitPrice       39
dtype: int64

Find all the customer emails that start with "J" and are from gmail.com

In [31]:
customers

CustomerId  FirstName      LastName  \
0            1       Luís     Gonçalves   
1            2     Leonie        Köhler   
2            3   François      Tremblay   
3            4      Bjørn        Hansen   
4            5  František   Wichterlová   
5            6     Helena          Holý   
6            7     Astrid        Gruber   
7            8       Daan       Peeters   
8            9       Kara       Nielsen   
9           10    Eduardo       Martins   
10          11  Alexandre         Rocha   
11          12    Roberto       Almeida   
12          13   Fernanda         Ramos   
13          14       Mark       Philips   
14          15   Jennifer      Peterson   
15          16      Frank        Harris   
16          17       Jack         Smith   
17          18   Michelle        Brooks   
18          19        Tim         Goyer   
19          20        Dan        Miller   
20          21      Kathy         Chase   
21          22    Heather       Leacock   
22          23       John        Gordon   
23          24      Frank       Ralston   
24          25     Victor       Stevens   
25          26    Richard    Cunningham   
26          27    Patrick          Gray   
27          28      Julia       Barnett   
28          29     Robert         Brown   
29          30     Edward       Francis   
30          31     Martha          Silk   
31          32      Aaron      Mitchell   
32          33      Ellie      Sullivan   
33          34       João     Fernandes   
34          35   Madalena       Sampaio   
35          36     Hannah     Schneider   
36          37       Fynn    Zimmermann   
37          38     Niklas      Schröder   
38          39    Camille       Bernard   
39          40  Dominique      Lefebvre   
40          41       Marc        Dubois   
41          42      Wyatt        Girard   
42          43   Isabelle       Mercier   
43          44      Terhi    Hämäläinen   
44          45   Ladislav        Kovács   
45          46       Hugh      O'Reilly   
46          47      Lucas       Mancini   
47          48   Johannes  Van der Berg   
48          49  Stanisław        Wójcik   
49          50    Enrique         Muñoz   
50          51     Joakim     Johansson   
51          52       Emma         Jones   
52          53       Phil        Hughes   
53          54      Steve        Murray   
54          55       Mark        Taylor   
55          56      Diego     Gutiérrez   
56          57       Luis         Rojas   
57          58      Manoj        Pareek   
58          59       Puja    Srivastava   

                                             Company  \
0   Embraer - Empresa Brasileira de Aeronáutica S.A.   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                   JetBrains s.r.o.   
5                                                NaN   
6                                                NaN   
7                                                NaN   
8                                                NaN   
9                                   Woodstock Discos   
10                              Banco do Brasil S.A.   
11                                            Riotur   
12                                               NaN   
13                                             Telus   
14                                     Rogers Canada   
15                                       Google Inc.   
16                             Microsoft Corporation   
17                                               NaN   
18                                        Apple Inc.   
19                                               NaN   
20                                               NaN   
21                                               NaN   
22                                               NaN   
23                                               NaN   
24                     

In [32]:
customers[customers["Email"].str.contains("gmail.com")]

CustomerId  FirstName  LastName Company                Address  \
2            3   François  Tremblay     NaN      1498 rue Bélanger   
5            6     Helena      Holý     NaN          Rilská 3174/6   
21          22    Heather   Leacock     NaN       120 S Orange Ave   
23          24      Frank   Ralston     NaN  162 E Superior Street   
27          28      Julia   Barnett     NaN            302 S 700 E   
30          31     Martha      Silk     NaN  194A Chain Lake Drive   
39          40  Dominique  Lefebvre     NaN         8, Rue Hanovre   
52          53       Phil    Hughes     NaN           113 Lupus St   

              City State         Country PostalCode               Phone  Fax  \
2         Montréal    QC          Canada    H2G 1A7   +1 (514) 721-4711  NaN   
5           Prague   NaN  Czech Republic      14300    +420 2 4177 0449  NaN   
21         Orlando    FL             USA      32801   +1 (407) 999-7788  NaN   
23         Chicago    IL             USA      60611   +1 (312) 332-3232  NaN   
27  Salt Lake City    UT             USA      84102   +1 (801) 531-7272  NaN   
30         Halifax    NS          Canada    B3S 1C5   +1 (902) 450-0450  NaN   
39           Paris   NaN          France      75002  +33 01 47 42 71 71  NaN   
52          London   NaN  United Kingdom   SW1V 3EN   +44 020 7976 5722  NaN   

                          Email  SupportRepId  
2           ftremblay@gmail.com             3  
5               hholy@gmail.com             5  
21           hleacock@gmail.com             4  
23           fralston@gmail.com             3  
27          jubarnett@gmail.com             5  
30         marthasilk@gmail.com             5  
39  dominiquelefebvre@gmail.com             4  
52        phil.hughes@gmail.com             3

Enter the one email address returned (you will likely need to scroll to the right) below.

jubarnett@gmail.com

Find all the invoices from the billing city Brasília, Edmonton, and Vancouver and sort in descending order by invoice ID.

In [33]:
invoice

InvoiceId  CustomerId          InvoiceDate  \
0            1           2  2009-01-01 00:00:00   
1            2           4  2009-01-02 00:00:00   
2            3           8  2009-01-03 00:00:00   
3            4          14  2009-01-06 00:00:00   
4            5          23  2009-01-11 00:00:00   
..         ...         ...                  ...   
407        408          25  2013-12-05 00:00:00   
408        409          29  2013-12-06 00:00:00   
409        410          35  2013-12-09 00:00:00   
410        411          44  2013-12-14 00:00:00   
411        412          58  2013-12-22 00:00:00   

                               BillingAddress BillingCity BillingState  \
0                     Theodor-Heuss-Straße 34   Stuttgart          NaN   
1                            Ullevålsveien 14        Oslo          NaN   
2                             Grétrystraat 63    Brussels          NaN   
3                              8210 111 ST NW    Edmonton           AB   
4                             69 Salem Street      Boston           MA   
..                                        ...         ...          ...   
407                     319 N. Frances Street     Madison           WI   
408                    796 Dundas Street West     Toronto           ON   
409  Rua dos Campeões Europeus de Viena, 4350       Porto          NaN   
410                           Porthaninkatu 9    Helsinki          NaN   
411                       12,Community Centre       Delhi          NaN   

    BillingCountry BillingPostalCode  Total  
0          Germany             70174   1.98  
1           Norway              0171   3.96  
2          Belgium              1000   5.94  
3           Canada           T6G 2C7   8.91  
4              USA              2113  13.86  
..             ...               ...    ...  
407            USA             53703   3.96  
408         Canada           M6J 1V1   5.94  
409       Portugal               NaN   8.91  
410        Finland             00530  13.86  
411          India            110017   1.99  

[412 rows x 9 columns]

In [34]:
brasilia = invoice[invoice["BillingCity"] == "Brasília"]

In [35]:
brasilia

InvoiceId  CustomerId          InvoiceDate BillingAddress BillingCity  \
34          35          13  2009-06-05 00:00:00   Qe 7 Bloco G    Brasília   
57          58          13  2009-09-07 00:00:00   Qe 7 Bloco G    Brasília   
79          80          13  2009-12-10 00:00:00   Qe 7 Bloco G    Brasília   
131        132          13  2010-07-31 00:00:00   Qe 7 Bloco G    Brasília   
252        253          13  2012-01-22 00:00:00   Qe 7 Bloco G    Brasília   
263        264          13  2012-03-03 00:00:00   Qe 7 Bloco G    Brasília   
318        319          13  2012-11-01 00:00:00   Qe 7 Bloco G    Brasília   

    BillingState BillingCountry BillingPostalCode  Total  
34            DF         Brazil         71020-677   1.98  
57            DF         Brazil         71020-677   3.96  
79            DF         Brazil         71020-677   5.94  
131           DF         Brazil         71020-677   0.99  
252           DF         Brazil         71020-677   1.98  
263           DF         Brazil         71020-677  13.86  
318           DF         Brazil         71020-677   8.91

In [36]:
edmonton = invoice[invoice["BillingCity"] == "Edmonton"]

In [37]:
edmonton

InvoiceId  CustomerId          InvoiceDate  BillingAddress BillingCity  \
3            4          14  2009-01-06 00:00:00  8210 111 ST NW    Edmonton   
132        133          14  2010-08-13 00:00:00  8210 111 ST NW    Edmonton   
155        156          14  2010-11-15 00:00:00  8210 111 ST NW    Edmonton   
177        178          14  2011-02-17 00:00:00  8210 111 ST NW    Edmonton   
229        230          14  2011-10-08 00:00:00  8210 111 ST NW    Edmonton   
350        351          14  2013-03-31 00:00:00  8210 111 ST NW    Edmonton   
361        362          14  2013-05-11 00:00:00  8210 111 ST NW    Edmonton   

    BillingState BillingCountry BillingPostalCode  Total  
3             AB         Canada           T6G 2C7   8.91  
132           AB         Canada           T6G 2C7   1.98  
155           AB         Canada           T6G 2C7   3.96  
177           AB         Canada           T6G 2C7   5.94  
229           AB         Canada           T6G 2C7   0.99  
350           AB         Canada           T6G 2C7   1.98  
361           AB         Canada           T6G 2C7  13.86

In [38]:
vancouver = invoice[invoice["BillingCity"] == "Vancouver"]

In [39]:
vancouver

InvoiceId  CustomerId          InvoiceDate       BillingAddress  \
35          36          15  2009-06-05 00:00:00  700 W Pender Street   
46          47          15  2009-07-16 00:00:00  700 W Pender Street   
101        102          15  2010-03-16 00:00:00  700 W Pender Street   
230        231          15  2011-10-21 00:00:00  700 W Pender Street   
253        254          15  2012-01-23 00:00:00  700 W Pender Street   
275        276          15  2012-04-26 00:00:00  700 W Pender Street   
327        328          15  2012-12-15 00:00:00  700 W Pender Street   

    BillingCity BillingState BillingCountry BillingPostalCode  Total  
35    Vancouver           BC         Canada           V6C 1G8   1.98  
46    Vancouver           BC         Canada           V6C 1G8  13.86  
101   Vancouver           BC         Canada           V6C 1G8   9.91  
230   Vancouver           BC         Canada           V6C 1G8   1.98  
253   Vancouver           BC         Canada           V6C 1G8   3.96  
275   Vancouver           BC         Canada           V6C 1G8   5.94  
327   Vancouver           BC         Canada           V6C 1G8   0.99

In [42]:
brasilia.sort_values(by="InvoiceId")

InvoiceId  CustomerId          InvoiceDate BillingAddress BillingCity  \
34          35          13  2009-06-05 00:00:00   Qe 7 Bloco G    Brasília   
57          58          13  2009-09-07 00:00:00   Qe 7 Bloco G    Brasília   
79          80          13  2009-12-10 00:00:00   Qe 7 Bloco G    Brasília   
131        132          13  2010-07-31 00:00:00   Qe 7 Bloco G    Brasília   
252        253          13  2012-01-22 00:00:00   Qe 7 Bloco G    Brasília   
263        264          13  2012-03-03 00:00:00   Qe 7 Bloco G    Brasília   
318        319          13  2012-11-01 00:00:00   Qe 7 Bloco G    Brasília   

    BillingState BillingCountry BillingPostalCode  Total  
34            DF         Brazil         71020-677   1.98  
57            DF         Brazil         71020-677   3.96  
79            DF         Brazil         71020-677   5.94  
131           DF         Brazil         71020-677   0.99  
252           DF         Brazil         71020-677   1.98  
263           DF         Brazil         71020-677  13.86  
318           DF         Brazil         71020-677   8.91

What is the total invoice amount of the first record returned? Enter the number below without a $ sign. Remember to sort in descending order to get the correct answer.

1.98

Show the number of orders placed by each customer (hint: this is found in the invoices table) and sort the result by the number of orders in descending order.

In [44]:
invoice

InvoiceId  CustomerId          InvoiceDate  \
0            1           2  2009-01-01 00:00:00   
1            2           4  2009-01-02 00:00:00   
2            3           8  2009-01-03 00:00:00   
3            4          14  2009-01-06 00:00:00   
4            5          23  2009-01-11 00:00:00   
..         ...         ...                  ...   
407        408          25  2013-12-05 00:00:00   
408        409          29  2013-12-06 00:00:00   
409        410          35  2013-12-09 00:00:00   
410        411          44  2013-12-14 00:00:00   
411        412          58  2013-12-22 00:00:00   

                               BillingAddress BillingCity BillingState  \
0                     Theodor-Heuss-Straße 34   Stuttgart          NaN   
1                            Ullevålsveien 14        Oslo          NaN   
2                             Grétrystraat 63    Brussels          NaN   
3                              8210 111 ST NW    Edmonton           AB   
4                             69 Salem Street      Boston           MA   
..                                        ...         ...          ...   
407                     319 N. Frances Street     Madison           WI   
408                    796 Dundas Street West     Toronto           ON   
409  Rua dos Campeões Europeus de Viena, 4350       Porto          NaN   
410                           Porthaninkatu 9    Helsinki          NaN   
411                       12,Community Centre       Delhi          NaN   

    BillingCountry BillingPostalCode  Total  
0          Germany             70174   1.98  
1           Norway              0171   3.96  
2          Belgium              1000   5.94  
3           Canada           T6G 2C7   8.91  
4              USA              2113  13.86  
..             ...               ...    ...  
407            USA             53703   3.96  
408         Canada           M6J 1V1   5.94  
409       Portugal               NaN   8.91  
410        Finland             00530  13.86  
411          India            110017   1.99  

[412 rows x 9 columns]

In [45]:
invoice.groupby("CustomerId").count()

InvoiceId  InvoiceDate  BillingAddress  BillingCity  BillingState  \
CustomerId                                                                      
1                   7            7               7            7             7   
2                   7            7               7            7             0   
3                   7            7               7            7             7   
4                   7            7               7            7             0   
5                   7            7               7            7             0   
6                   7            7               7            7             0   
7                   7            7               7            7             0   
8                   7            7               7            7             0   
9                   7            7               7            7             0   
10                  7            7               7            7             7   
11                  7            7               7            7             7   
12                  7            7               7            7             7   
13                  7            7               7            7             7   
14                  7            7               7            7             7   
15                  7            7               7            7             7   
16                  7            7               7            7             7   
17                  7            7               7            7             7   
18                  7            7               7            7             7   
19                  7            7               7            7             7   
20                  7            7               7            7             7   
21                  7            7               7            7             7   
22                  7            7               7            7             7   
23                  7            7               7            7             7   
24                  7            7               7            7             7   
25                  7            7               7            7             7   
26                  7            7               7            7             7   
27                  7            7               7            7             7   
28                  7            7               7            7             7   
29                  7            7               7            7             7   
30                  7            7               7            7             7   
31                  7            7               7            7             7   
32                  7            7               7            7             7   
33                  7            7               7            7             7   
34                  7            7               7            7             0   
35                  7            7               7            7             0   
36                  7            7               7            7             0   
37                  7            7               7            7             0   
38                  7            7               7            7             0   
39                  7            7               7            7             0   
40                  7            7               7            7             0   
41                  7            7               7            7             0   
42                  7            7               7            7             0   
43                  7            7               7            7             0   
44                  7            7               7            7             0   
45                  7            7               7            7             0   
46                  7            7               7            7             7   
47                  7            7               7            7             7   
48                  7            7         

What is the number of items placed for the 8th person on this list? Enter that number below.

7

Find the albums with 12 or more tracks

In [46]:
tracks

TrackId                                               Name  AlbumId  \
0           1            For Those About To Rock (We Salute You)        1   
1           2                                  Balls to the Wall        2   
2           3                                    Fast As a Shark        3   
3           4                                  Restless and Wild        3   
4           5                               Princess of the Dawn        3   
...       ...                                                ...      ...   
3498     3499  Pini Di Roma (Pinien Von Rom) \ I Pini Della V...      343   
3499     3500  String Quartet No. 12 in C Minor, D. 703 "Quar...      344   
3500     3501               L'orfeo, Act 3, Sinfonia (Orchestra)      345   
3501     3502  Quintet for Horn, Violin, 2 Violas, and Cello ...      346   
3502     3503                                      Koyaanisqatsi      347   

      MediaTypeId  GenreId                                           Composer  \
0               1        1          Angus Young, Malcolm Young, Brian Johnson   
1               2        1                                                NaN   
2               2        1  F. Baltes, S. Kaufman, U. Dirkscneider & W. Ho...   
3               2        1  F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. D...   
4               2        1                         Deaffy & R.A. Smith-Diesel   
...           ...      ...                                                ...   
3498            2       24                                                NaN   
3499            2       24                                     Franz Schubert   
3500            2       24                                 Claudio Monteverdi   
3501            2       24                            Wolfgang Amadeus Mozart   
3502            2       10                                       Philip Glass   

      Milliseconds     Bytes  UnitPrice  
0           343719  11170334       0.99  
1           342562   5510424       0.99  
2           230619   3990994       0.99  
3           252051   4331779       0.99  
4           375418   6290521       0.99  
...            ...       ...        ...  
3498        286741   4718950       0.99  
3499        139200   2283131       0.99  
3500         66639   1189062       0.99  
3501        221331   3665114       0.99  
3502        206005   3305164       0.99  

[3503 rows x 9 columns]

In [62]:
tracks.groupby("Name").count()

TrackId  AlbumId  \
Name                                                                   
"40"                                                      1        1   
"?"                                                       1        1   
"Eine Kleine Nachtmusik" Serenade In G, K. 525:...        1        1   
#1 Zero                                                   1        1   
#9 Dream                                                  1        1   
...                                                     ...      ...   
É que Nessa Encarnação Eu Nasci Manga                     1        1   
Étude 1, In C Major - Preludio (Presto) - Liszt           1        1   
Óculos                                                    1        1   
Óia Eu Aqui De Novo                                       1        1   
Último Pau-De-Arara                                       1        1   

                                                    MediaTypeId  GenreId  \
Name                                                                       
"40"                                                          1        1   
"?"                                                           1        1   
"Eine Kleine Nachtmusik" Serenade In G, K. 525:...            1        1   
#1 Zero                                                       1        1   
#9 Dream                                                      1        1   
...                                                         ...      ...   
É que Nessa Encarnação Eu Nasci Manga                         1        1   
Étude 1, In C Major - Preludio (Presto) - Liszt               1        1   
Óculos                                                        1        1   
Óia Eu Aqui De Novo                                           1        1   
Último Pau-De-Arara                                           1        1   

                                                    Composer  Milliseconds  \
Name                                                                         
"40"                                                       1             1   
"?"                                                        0             1   
"Eine Kleine Nachtmusik" Serenade In G, K. 525:...         1             1   
#1 Zero                                                    1             1   
#9 Dream                                                   0             1   
...                                                      ...           ...   
É que Nessa Encarnação Eu Nasci Manga                      1             1   
Étude 1, In C Major - Preludio (Presto) - Liszt            0             1   
Óculos                                                     0             1   
Óia Eu Aqui De Novo                                        0             1   
Último Pau-De-Arara                                        1             1   

                                                    Bytes  UnitPrice  
Name                                                                  
"40"                                                    1          1  
"?"                                                     1          1  
"Eine Kleine Nachtmusik" Serenade In G, K. 525:...      1          1  
#1 Zero                                                 1          1  
#9 Dream                                                1          1  
...                                                   ...        ...  
É que Nessa Encarnação Eu Nasci Manga                   1          1  
Étude 1, In C Major - Preludio (Presto) - Liszt         1          1  
Óculos                                                  1          1  
Óia Eu Aqui De Novo                                     1          1  
Último Pau-De-Arara                                     1          1  

[3257 rows x 8 columns]

In [63]:
tracks_df = pd.DataFrame(tracks.groupby("AlbumId")["TrackId"].count())

In [64]:
tracks_df

TrackId
AlbumId         
1             10
2              1
3              3
4              8
5             15
...          ...
343            1
344            1
345            1
346            1
347            1

[347 rows x 1 columns]

In [65]:
tracks_df.sort_values(by="TrackId", ascending=False)

TrackId
AlbumId         
141           57
23            34
73            30
229           26
251           25
...          ...
289            1
288            1
287            1
286            1
347            1

[347 rows x 1 columns]

In [66]:
tracks_df[tracks_df["TrackId"]>=12]

TrackId
AlbumId         
5             15
6             13
7             12
8             14
10            14
...          ...
259           17
261           17
270           14
271           14
321           12

[158 rows x 1 columns]

While the number of records returned is limited to 10, the query, if run correctly, will indicate how many total records there are. Enter that number below.

158